pkg/storegateway/sharding_strategy_test.go

```diff
@@ -499,3 +500,345 @@ type shardingLimitsMock struct {
+	assertStateUnchanged := func(t *testing.T, baseline, current map[string]instanceState, instanceIDs []string) {
+		t.Helper()
+		for _, id := range instanceIDs {
+			baselineState, ok := baseline[id]
+			require.True(t, ok, "baseline state not found for instance %s", id)
+
+			currentState, ok := current[id]
+			require.True(t, ok, "current state not found for instance %s", id)
+
+			assert.ElementsMatch(t, baselineState.users, currentState.users,
+				"FilterUsers result changed for instance %s", id)
+
+			// Check blocks per user to ensure no redistribution within users.
+			require.Equal(t, len(baselineState.blocksPerUser), len(currentState.blocksPerUser),
+				"Number of users with blocks changed for instance %s", id)
+
+			for userID, baselineBlocks := range baselineState.blocksPerUser {
+				currentBlocks, ok := currentState.blocksPerUser[userID]
+				require.True(t, ok, "user %s blocks not found in current state for instance %s", userID, id)
+				assert.ElementsMatch(t, baselineBlocks, currentBlocks,
+					"FilterBlocks result changed for instance %s, user %s", id, userID)
+			}
+		}
+	}
+
+	// Create KV store.
+	store, closer := consul.NewInMemoryClient(ring.GetCodec(), log.NewNopLogger(), nil)
+	t.Cleanup(func() { assert.NoError(t, closer.Close()) })
+
+	limits := &shardingLimitsMock{storeGatewayTenantShardSize: shardSize}
+
+	// Helper to add instances to ring descriptor.
+	addInstancesToRing := func(d *ring.Desc, instances []instanceDef) {
+		for _, inst := range instances {
+			d.AddIngester(inst.id, inst.addr, inst.zone, inst.tokens, ring.ACTIVE, registeredAt, false, time.Time{}, nil)
+		}
+	}
+
+	// Helper to remove instances from ring descriptor.
+	removeInstancesFromRing := func(d *ring.Desc, instances []instanceDef) {
+		for _, inst := range instances {
+			d.RemoveIngester(inst.id)
+		}
+	}
+
+	// Helper to get instance IDs from definitions.
+	getInstanceIDs := func(instances ...[]instanceDef) []string {
+		var ids []string
+		for _, instList := range instances {
+			for _, inst := range instList {
+				ids = append(ids, inst.id)
+			}
+		}
+		return ids
+	}
+
+	// ============================================================================
+	// Step 1: Initial state - 3 zones, RF=3
+	// ============================================================================
+	t.Log("Step 1: Setting up initial ring with 3 zones, RF=3")
+
+	require.NoError(t, store.CAS(ctx, "test", func(interface{}) (interface{}, bool, error) {
+		d := ring.NewDesc()
+		addInstancesToRing(d, zoneAInstances)
+		addInstancesToRing(d, zoneBInstances)
+		addInstancesToRing(d, zoneCInstances)
+		return d, true, nil
+	}))
+
+	ringCfg := ring.Config{
+		ReplicationFactor:    3,
+		HeartbeatTimeout:     time.Minute,
+		SubringCacheDisabled: true,
+		ZoneAwarenessEnabled: true,
+	}
+
+	r, err := ring.NewWithStoreClientAndStrategy(ringCfg, "test", "test", store, ring.NewIgnoreUnhealthyInstancesReplicationStrategy(), nil, log.NewNopLogger())
```

Comment:
[nit] We can do like store-gateway does:

```suggestion
	r, err := ring.NewWithStoreClientAndStrategy(ringCfg, RingNameForServer, RingKey, store, ring.NewIgnoreUnhealthyInstancesReplicationStrategy(), nil, log.NewNopLogger())
```